src/clj/rems/workflow/dynamic.clj

```diff
@@ -35,12 +29,154 @@
-(def EventTypes #{:event/draft-saved :event/submitted :event/returned #_:event/license-required #_:event/license-accepted :event/comment-requested :event/commented :event/decision-requested :event/decided :event/approved :event/rejected :event/closed :event/member-added})
-
+(def States
+  #{::approved
+    ::closed
+    ::draft
+    ::rejected
+    ::submitted})
+(def CommandTypes
+  #{#_::accept-license
+    #_::require-license
+    ::add-member
+    ::approve
+    ::close
+    ::comment
+    ::decide
+    ::reject
+    ::request-comment
+    ::request-decision
+    ::return
+    ::save-draft
+    ::submit})
+(def EventTypes
+  #{#_:event/license-accepted
+    #_:event/license-required
+    :event/approved
+    :event/closed
+    :event/comment-requested
+    :event/commented
+    :event/decided
+    :event/decision-requested
+    :event/draft-saved
+    :event/member-added
+    :event/rejected
+    :event/returned
+    :event/submitted})
+
+;; TODO: namespaced keys e.g. :event/type, :event/time, :event/actor, :application/id
+;; TODO: add version number to events
+(s/defschema EventBase
+  {:event (apply s/enum EventTypes)
+   :application-id s/Int
+   :actor UserId
+   :time DateTime})
 
+(s/defschema ApprovedEvent
+  (assoc EventBase
+         :event (s/eq :event/approved)
+         :comment s/Str))
+(s/defschema ClosedEvent
+  (assoc EventBase
+         :event (s/eq :event/closed)
+         :comment s/Str))
+(s/defschema CommentedEvent
+  (assoc EventBase
+         :event (s/eq :event/commented)
+         :comment s/Str))
+(s/defschema CommentRequestedEvent
+  (assoc EventBase
+         :event (s/eq :event/comment-requested)
+         :commenters [s/Str]
+         :comment s/Str))
+(s/defschema DecidedEvent
+  (assoc EventBase
+         :event (s/eq :event/decided)
+         :decision (s/enum :approved :rejected)
+         :comment s/Str))
+(s/defschema DecisionRequestedEvent
+  (assoc EventBase
+         :event (s/eq :event/decision-requested)
+         :decider s/Str
+         :comment s/Str))
+(s/defschema DraftSavedEvent
+  (assoc EventBase
+         :event (s/eq :event/draft-saved)
+         :items {Long s/Str}
+         :licenses {Long s/Str}))
+(s/defschema MemberAddedEvent
+  (assoc EventBase
+         :event (s/eq :event/member-added)
+         :member s/Str))
+(s/defschema RejectedEvent
+  (assoc EventBase
+         :event (s/eq :event/rejected)
+         :comment s/Str))
+(s/defschema ReturnedEvent
+  (assoc EventBase
+         :event (s/eq :event/returned)
+         :comment s/Str))
+(s/defschema SubmittedEvent
+  (assoc EventBase
+         :event (s/eq :event/submitted)))
+
+(def event-schemas
+  [ApprovedEvent
+   ClosedEvent
+   CommentedEvent
+   CommentRequestedEvent
+   DecidedEvent
+   DecisionRequestedEvent
+   DraftSavedEvent
+   MemberAddedEvent
+   RejectedEvent
+   ReturnedEvent
+   SubmittedEvent])
```

Comment:
since you anyway need to define a collection of event-schemas, why not
```
{:event/approved ApprovedEvent
 ...}
```
that way `get-event-type` wouldn't be needed

Reply:
and also EventTypes wouldn't be needed as it would just be `(keys event-schemas)`